File: src/main/java/com/vandendaelen/nicephore/client/gui/SettingsScreen.java
package com.vandendaelen.nicephore.client.gui;

import com.mojang.blaze3d.vertex.PoseStack;
import com.vandendaelen.nicephore.config.NicephoreConfig;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public class SettingsScreen extends Screen {
    private static final Component TITLE = Component.translatable("nicephore.gui.settings");

    protected SettingsScreen() {
        super(TITLE);
    }

    @Override
    public void render(@NotNull PoseStack pPoseStack, int pMouseX, int pMouseY, float pPartialTick) {
        this.renderBackground(pPoseStack);
        final int centerX = this.width / 2;
        final int startingLine = this.width / 2 - 150;

        drawCenteredString(pPoseStack, Minecraft.getInstance().font, Component.translatable("nicephore.gui.settings"), centerX, 35, Color.WHITE.getRGB());

        this.clearWidgets();
        this.addRenderableWidget(Button.builder(Component.translatable("nicephore.screenshot.exit"), button -> onClose()).bounds(this.width - 60, 10, 50, 20).build());
        this.addRenderableWidget(Button.builder(Component.translatable("nicephore.screenshot.showOptimisationStatus", NicephoreConfig.Client.getShouldShowOptStatus() ? "ON" : "OFF"), button -> changeShowOptimisationStatus(!NicephoreConfig.Client.getShouldShowOptStatus())).bounds(startingLine, 60, 300, 20).build());
        this.addRenderableWidget(Button.builder(Component.translatable("nicephore.screenshot.makeJPEGs", NicephoreConfig.Client.getJPEGToggle() ? "ON" : "OFF"), button -> changeMakeJPEGs(!NicephoreConfig.Client.getJPEGToggle())).bounds(startingLine, 90, 300, 20).build());
        this.addRenderableWidget(Button.builder(Component.translatable("nicephore.screenshot.screenshotCustomMessage", NicephoreConfig.Client.getScreenshotCustomMessage() ? "ON" : "OFF"), button -> changeScreenshotCustomMessage(!NicephoreConfig.Client.getScreenshotCustomMessage())).bounds(startingLine, 120, 300, 20).build());
        this.addRenderableWidget(Button.builder(Component.translatable("nicephore.screenshot.setScreenshotToClipboard", NicephoreConfig.Client.getScreenshotToClipboard() ? "ON" : "OFF"), button -> changeScreenshotToClipboard(!NicephoreConfig.Client.getScreenshotToClipboard())).bounds(startingLine, 150, 300, 20).build());

        super.render(pPoseStack, pMouseX, pMouseY, pPartialTick);
    }

    private void changeShowOptimisationStatus(boolean value) {
        NicephoreConfig.Client.setShouldShowOptStatus(value);
    }

    private void changeMakeJPEGs(boolean value) {
        NicephoreConfig.Client.setJPEGToggle(value);
    }

    private void changeScreenshotCustomMessage(boolean value) {
        NicephoreConfig.Client.setScreenshotCustomMessage(value);
    }

    private void changeScreenshotToClipboard(boolean value) {
        NicephoreConfig.Client.setScreenshotToClipboard(value);
    }

//    private void changeFilter() {
//        ScreenshotFilter nextFilter = NicephoreConfig.Client.getScreenshotFilter().next();
//        NicephoreConfig.Client.setScreenshotFilter(nextFilter);
//        init();
//
//        if (listener != null) {
//            listener.onFilterChange(nextFilter);
//        }
//    }
}

File: src/main/kotlin/com/vandendaelen/nicephore/client/gui/SettingsScreen.kt
package com.vandendaelen.nicephore.client.gui

import com.vandendaelen.nicephore.config.NicephoreConfig
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.network.chat.Component
import java.awt.Color

class SettingsScreen(private val onSettingsClosed: () -> Unit = {}) : AbstractNicephoreScreen(TITLE) {

    private var titleY: Int = 35

    override fun init() {
        super.init()
        refreshWidgets()
    }

    override fun buildWidgets() {
        val startingLine = this.width / 2 - 150
        val itemCount = 5
        val itemHeight = BUTTON_HEIGHT + PADDING  // 30
        val totalContentHeight = itemCount * itemHeight - PADDING  // 140
        val titleHeight = 9 + PADDING  // font height + gap
        val topY = ((this.height - titleHeight - totalContentHeight) / 2).coerceAtLeast(TOOLBAR_HEIGHT)
        titleY = topY
        val contentStartY = topY + titleHeight

        this.addRenderableWidget(
            Button.builder(Component.translatable("nicephore.screenshot.exit")) { onClose() }
                .bounds(this.width - PADDING - 50, PADDING, 50, BUTTON_HEIGHT).build()
        )
        this.addRenderableWidget(
            Button.builder(
                Component.translatable("nicephore.screenshot.showOptimisationStatus", if (NicephoreConfig.Client.getShouldShowOptStatus()) "ON" else "OFF")
            ) { toggleSetting { NicephoreConfig.Client.setShouldShowOptStatus(!NicephoreConfig.Client.getShouldShowOptStatus()) } }
                .bounds(startingLine, contentStartY, 300, BUTTON_HEIGHT).build()
        )
        this.addRenderableWidget(
            Button.builder(
                Component.translatable("nicephore.screenshot.makeJPEGs", if (NicephoreConfig.Client.getJPEGToggle()) "ON" else "OFF")
            ) { toggleSetting { NicephoreConfig.Client.setJPEGToggle(!NicephoreConfig.Client.getJPEGToggle()) } }
                .bounds(startingLine, contentStartY + itemHeight, 300, BUTTON_HEIGHT).build()
        )
        this.addRenderableWidget(
            Button.builder(
                Component.translatable("nicephore.screenshot.screenshotCustomMessage", if (NicephoreConfig.Client.getScreenshotCustomMessage()) "ON" else "OFF")
            ) { toggleSetting { NicephoreConfig.Client.setScreenshotCustomMessage(!NicephoreConfig.Client.getScreenshotCustomMessage()) } }
                .bounds(startingLine, contentStartY + 2 * itemHeight, 300, BUTTON_HEIGHT).build()
        )
        this.addRenderableWidget(
            Button.builder(
                Component.translatable("nicephore.screenshot.setScreenshotToClipboard", if (NicephoreConfig.Client.getScreenshotToClipboard()) "ON" else "OFF")
            ) { toggleSetting { NicephoreConfig.Client.setScreenshotToClipboard(!NicephoreConfig.Client.getScreenshotToClipboard()) } }
                .bounds(startingLine, contentStartY + 3 * itemHeight, 300, BUTTON_HEIGHT).build()
        )

        val currentColumns = NicephoreConfig.Client.getGalleryColumns()
        val label = if (currentColumns == 0) "Auto" else "$currentColumns"
        this.addRenderableWidget(
            Button.builder(
                Component.translatable("nicephore.settings.galleryColumns", label)
            ) { cycleGalleryColumns() }
                .bounds(startingLine, contentStartY + 4 * itemHeight, 300, BUTTON_HEIGHT).build()
        )
    }

    override fun extractRenderState(guiGraphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, partialTicks: Float) {
        val centerX = this.width / 2

        guiGraphics.centeredText(
            Minecraft.getInstance().font,
            Component.translatable("nicephore.gui.settings"),
            centerX, titleY, Color.WHITE.rgb
        )

        super.extractRenderState(guiGraphics, mouseX, mouseY, partialTicks)
    }

    private fun toggleSetting(action: () -> Unit) {
        action()
        refreshWidgets()
    }

    private fun cycleGalleryColumns() {
        val current = NicephoreConfig.Client.getGalleryColumns()
        val next = when (current) {
            0 -> 2
            in 2..5 -> current + 1
            else -> 0
        }
        NicephoreConfig.Client.setGalleryColumns(next)
        refreshWidgets()
    }

    override fun onClose() {
        super.onClose()
        onSettingsClosed()
    }

    companion object {
        private val TITLE = Component.translatable("nicephore.gui.settings")
    }
}
